### ISSUE TITLE
pi-manage create_enckey fails with old enum package

### ISSUE THREAD

**Issue by fberlakovich** [privacyidea/privacyidea#977]
# What did you try to do?

I followed the tutorial in http://privacyidea.readthedocs.io/en/latest/installation/pip.html and tried to create the encryption key with `pi-manage create_enckey`.

# What outcome did you expect?

I expected the encryption key to be created without an error.

# What outcome did you experience?

The command failed with the following error 
```
pi-manage create_enckey                                                                                                                                                                                                                           127 ↵
Traceback (most recent call last):
  File "/opt/privacyidea/bin/pi-manage", line 63, in <module>
    from privacyidea.lib.auth import (create_db_admin, list_db_admin,
  File "/opt/privacyidea/local/lib/python2.7/site-packages/privacyidea/lib/auth.py", line 25, in <module>
    from privacyidea.lib.token import check_user_pass
  File "/opt/privacyidea/local/lib/python2.7/site-packages/privacyidea/lib/token.py", line 72, in <module>
    from privacyidea.lib.tokenclass import TokenClass
  File "/opt/privacyidea/local/lib/python2.7/site-packages/privacyidea/lib/tokenclass.py", line 88, in <module>
    from .config import (get_from_config, get_prepend_pin)
  File "/opt/privacyidea/local/lib/python2.7/site-packages/privacyidea/lib/config.py", line 48, in <module>
    from .caconnectors.localca import BaseCAConnector
  File "/opt/privacyidea/local/lib/python2.7/site-packages/privacyidea/lib/caconnectors/localca.py", line 33, in <module>
    from OpenSSL import crypto
  File "/opt/privacyidea/local/lib/python2.7/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL
  File "/opt/privacyidea/local/lib/python2.7/site-packages/OpenSSL/crypto.py", line 12, in <module>
    from cryptography import x509
  File "/opt/privacyidea/local/lib/python2.7/site-packages/cryptography/x509/__init__.py", line 8, in <module>
    from cryptography.x509.base import (
  File "/opt/privacyidea/local/lib/python2.7/site-packages/cryptography/x509/base.py", line 16, in <module>
    from cryptography.x509.extensions import Extension, ExtensionType
  File "/opt/privacyidea/local/lib/python2.7/site-packages/cryptography/x509/extensions.py", line 24, in <module>
    from cryptography.x509.general_name import GeneralName, IPAddress, OtherName
  File "/opt/privacyidea/local/lib/python2.7/site-packages/cryptography/x509/general_name.py", line 18, in <module>
    from cryptography.x509.name import Name
  File "/opt/privacyidea/local/lib/python2.7/site-packages/cryptography/x509/name.py", line 28, in <module>
    _ASN1_TYPE_TO_ENUM = dict((i.value, i) for i in _ASN1Type)
TypeError: 'type' object is not iterable
```

# Configuration

* privacyIDEA Version: 2.21.4
```
pip show privacyidea                                                                                                                                                                                                                                1 ↵
---
Name: privacyIDEA
Version: 2.21.4
Location: /opt/privacyidea/lib/python2.7/site-packages
Requires: Flask, Flask-Migrate, Flask-SQLAlchemy, Flask-Script, Jinja2, Mako, MarkupSafe, PyMySQL, Pillow, PyJWT, PyYAML, Pygments, SQLAlchemy, Werkzeug, alembic, argparse, bcrypt, beautifulsoup4, cffi, configobj, docutils, funcparserlib, itsdangerous, ldap3, netaddr, passlib, pyasn1, pyOpenSSL, pycparser, pycrypto, pyrad, pyusb, qrcode, requests, sqlsoup, ecdsa, lxml, python-gnupg, defusedxml, flask-babe
```
* Installation method: Python virtualenv

## more details: 

* OS: Debian 8.9

# Resolution
It turned out I have this exact issue: https://github.com/pyca/cryptography/issues/4010
Removing the `enum` package_
```
pip uninstall enum
Uninstalling enum:
  /opt/privacyidea/lib/python2.7/site-packages/enum-0.4.6-py2.7.egg-info
  /opt/privacyidea/lib/python2.7/site-packages/enum.py
  /opt/privacyidea/lib/python2.7/site-packages/enum.pyc
```
and installing the ```enum34``` package:
```
pip install enum34
Downloading/unpacking enum34
  Downloading enum34-1.1.6-py2-none-any.whl
Installing collected packages: enum34
Successfully installed enum34
```
fixed the issue for me. However, maybe it would be possible to catch this error in `pi-manage create_enckey` and give brief instructions on how to fix it.

**Comment by cornelinux** [comment]
Thanks a lot for the report and the solution! :-)
This is indeed annoying, since the error occurs in the CA connector. Probably 98% are not using this code, We will figure something out!

**Comment by cornelinux** [comment]
Interesting. The python virtual environment on your debian system seems to install other stuff than the virtual env on my ubuntu:

    Collecting enum34; python_version < "3" (from cryptography>=2.1.4->pyOpenSSL>=0.15.1>privacyidea)
       Using cached enum34-1.1.6-py2-none-any.whl

Actually in my case ``enum34`` is installed right from the start!

Which version of ``virtualenv`` are you using?

Maybe you have an older version, where you still need to use: ``--no-site-packages``?

    virtualenv --no-site-packages /opt/privacyidea

Can you please check this?

**Comment by fberlakovich** [comment]
> Can you please check this?

Yes, sure!
```
virtualenv --version
1.11.6
```
I tried the whole process again in a fresh virtualenv and it seems like I hat to install the packages manually when I set up privacyidea (I did it last week, but couldn't finish the installation back then). If i run `pi-manage create_enckey` in the new virtualenv it lacks the `enum` package completely, as well as several other packages:
```
pi-manage create_enckey
<output omitted>
ImportError: No module named enum
```
According to my history I installed all the missing packages, including `enum`. 

PS: I also had to upgrade my `setuptools` in order to install privacyidea. This might be worth noting in the documentation. Should I create another issue for that?

**Comment by cornelinux** [comment]
No need to create a new issue for setup tools. We will add the according note in the setup:

1. within virutalenv run "pip install --upgrade setuptools".

In regards to the enum this really looks wired. I also saw that I also at the moment had to install the modules "enum34" **and** "ipaddress", which both were missing in the dependencies in cryptography.

Honestly currently I am not sure if I want to add this either to the default documentation or to the dependencies of privacyIDEA, since these are no direct dependencies. 

Anyhow this (at the time of writing) works like a charm:

    virtualenv /opt/privacyidea
    source /opt/privacyidea/bin/activate
    pip install --upgrade setuptools
    pip install enum34
    pip install ipaddress
    pip install privacyidea
    pi-manage create_enckey

**Comment by fberlakovich** [comment]
> Anyhow this (at the time of writing) works like a charm:

This sequence of commands also works for me.

**Comment by cornelinux** [comment]
Yes, currently I am a bit clueless if or how we should add this to the docs. I suspect in a few weeks these "fixes" will not be necessary anymore.

**Comment by fberlakovich** [comment]
> Yes, currently I am a bit clueless if or how we should add this to the docs. I suspect in a few weeks these "fixes" will not be necessary anymore.

Is there something like a knowledge base or "known issues"?. Instead of including a resolution in the documentation that is only applicable to a small usergroup, there could be a link to a general page describing issues found. If a user is affected by an issue during the installation they would most likely visit this page.

**Comment by cornelinux** [comment]
We could add another category at https://community.privacyidea.org/c/installation.
Somehting like "know issues" or "knowledge base".
There is a category installation - but usually these are questions about installation.

But we might also add a *Tag* "known issues" and add an information like this to the *category* "installation" with the *tag* "known issue". 

It is always a difficult decision how to organize information and data ;-)

**Comment by fberlakovich** [comment]
> But we might also add a Tag "known issues" and add an information like this to the category "installation" with the tag "known issue".

Personally I like this variant best. A tag "known issues" might possibly be useful in other categories too. 

> It is always a difficult decision how to organize information and data ;-)

It is indeed! But as long as the "known issues" tag/category is easier to find than manually searching for similar issues in Google, it is already an improvement ;).

**Comment by cornelinux** [comment]
I added a tag "known issue". So if you want to, you can describe this at community.privacyidea.org :-)
